File: src/main/java/yaasl/server/security/PasswordEncoder.java
package yaasl.server.security;

import org.apache.commons.codec.binary.Hex;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Component;

import java.security.MessageDigest;

import static org.apache.commons.codec.binary.Hex.encodeHexString;

@Component
public class PasswordEncoder implements org.springframework.security.crypto.password.PasswordEncoder {

    private BCryptPasswordEncoder bCryptPasswordEncoder = new BCryptPasswordEncoder();
    private MessageDigest md5Digest;

    public PasswordEncoder() throws Exception {
        this.md5Digest = MessageDigest.getInstance("md5");
    }

    @Override
    public String encode(CharSequence rawPassword) {
        return bCryptPasswordEncoder.encode(rawPassword);
    }

    @Override
    public boolean matches(CharSequence rawPassword, String encodedPassword) {
        if (bCryptPasswordEncoder.matches(rawPassword, encodedPassword)) {
            return true;
        }
        else {
            return encodedPassword.equals(encodeHexString(md5Digest.digest(rawPassword.toString().getBytes())));
        }
    }

}

File: src/main/kotlin/yaasl/server/security/PasswordEncoder.kt
package yaasl.server.security

import org.apache.commons.codec.binary.Hex.encodeHexString
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Component
import java.security.MessageDigest

@Component
class PasswordEncoder: org.springframework.security.crypto.password.PasswordEncoder {

    private val bCryptPasswordEncoder = BCryptPasswordEncoder()
    private val md5Digest: MessageDigest

    init {
        this.md5Digest = MessageDigest.getInstance("md5")
    }

    override fun encode(rawPassword: CharSequence): String {
        return bCryptPasswordEncoder.encode(rawPassword)
    }

    override fun matches(rawPassword: CharSequence, encodedPassword: String): Boolean {
        return if (bCryptPasswordEncoder.matches(rawPassword, encodedPassword)) {
            true
        } else {
            encodedPassword == encodeHexString(md5Digest.digest(rawPassword.toString().toByteArray()))
        }
    }

}
